Client/Frontend/Browser/TabScrollController.swift

```diff
@@ -198,18 +199,35 @@ private extension TabScrollingController {
         return y
     }
 
-    func animateToolbarsWithOffsets(animated animated: Bool, duration: NSTimeInterval, headerOffset: CGFloat,
+    func animateToolbarsWithOffsets(animated animated: Bool, duration: NSTimeInterval, headerOffset: CGFloat, scrollOffset: CGFloat,
         footerOffset: CGFloat, alpha: CGFloat, completion: ((finished: Bool) -> Void)?) {
 
+        // Due to the header not being placed on top of the scroll view, when it animates away, and the scroll offset is negative,
+        // its superview background will show through temporarily, leading to anpleasant visual effect. By inserted a subview of the
+        // same colour as the background, we can hide this effect for the duration of the animation.
+        var veil: UIView?
+        if let header = self.header, superview = header.superview where animated {
```

Comment:
Yeah, this was a gnarly issue. I spent quite a while trying to figure out the best way to get rid of the problem.
That is the graphical issue I mention, yes. The problem is that previously, the web content was linked by a constraint to the URL bar, and so when the URL bar animated away, the web content would maintain its relative offset to the URL bar, meaning the bar didn't lift off the web content, and the black bar did not appear.
Now, the URL bar is independent of the content, with regards to its positioning. The problem is, that when the URL bar is lifted above the web content (for example, if you scroll up into the "elastic" region), when it animates away, because the web content does not maintain that relative offset, there is nothing behind the URL bar for a short period (while the scroll is bouncing back to an offset of 0), which causes the graphical issue.

A timing based approach would be difficult to pull off. For the graphical issue to be resolved with that approach, we would want to move the URL bar only when the web content starts pushing it. This would not be immediate, and would be variable depending on the amount the user has scrolled up above the page — this means we would need to start the URL bar animation at a different time to the main animation. What's more, the animation is not linear, so in order for it to line up precisely, we would need to start the URL bar animation at a different time, with a different timing function that depended on how far the user had scrolled up (or, equivalent, the velocity of the bounce).
The reason the timing of the bar is not the same as the timing of the web content is that it is always scrolling a constant amount, whereas everything else is scrolling based on the offset — which is why a perfect animation would only start the animation of the URL bar at the instant that the `contentOffset.y` was exactly (negative) the height of the URL bar.

Maybe I've overlooked something, but I spent quite a while trying to fix this in a neater way, and had trouble devising a solution that actually looked right... But did you mean something else when you suggested a velocity-based approach?